File: 230.kth-smallest-element-in-a-bst.py
#
# @lc app=leetcode id=230 lang=python3
#
# [230] Kth Smallest Element in a BST
#

# Time:  O(max(h, k))
# Space: O(h)


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        count, stack = 0, [(root, False)]
        while stack:
            node, is_visited = stack.pop()
            if node is None:
                continue
            if is_visited:
                count += 1
                if count == k:
                    return node.val
            else:
                stack.append((node.right, False))
                stack.append((node, True))
                stack.append((node.left, False))
        return 0

        node, stack, count = root, [], 0
        while node or stack:
            while node:
                stack.append(node)
                node = node.left
            node, count = stack.pop(), count + 1
            if count == k:
                return node.val
            node = node.right
        return 0


# @lc code=end

File: 230.kth-smallest-element-in-a-bst.js
/*
 * @lc app=leetcode id=230 lang=javascript
 *
 * [230] Kth Smallest Element in a BST
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {
  const stk = [];
  let node = root;
  let count = 0;
  while (node || stk.length) {
    while (node) {
      stk.push(node);
      node = node.left;
    }
    node = stk.pop();
    if (++count === k) {
      return node.val;
    }
    node = node.right;
  }
  return 0;
};
// @lc code=end
